### ISSUE TITLE
Nomenclature of aur(1) scripts

### ISSUE THREAD

**Issue by AladW** [AladW/aurutils#382]
Follow-up on https://github.com/AladW/aurutils/pull/311#issuecomment-383293423

I'm mostly happy with how the aurutils scripts are named, but there's some candidates where I have doubts:

- [X] `aur-arrow`, previously `aurqueue` and `aur-deps-src`. Since `tsort` was moved out from this script, it's no longer a "queue", and the other name didn't seem suitable either (for example, it's not related to `aur-deps-rpc`). `arrow` emphasizes on the paired output `pkgbase -> depends`, but says nothing about the actual goal of the script, being piping the output to `tsort` to get the build order in a directory of PKGBUILDs. (__renamed to `aur-graph`__)
- [x] `aur-deps-rpc`, previously `aurchain`. `chain` was a fancy name to indicate the recursive approach of querying the AUR, which is only about the implementation. `deps-rpc` is closer to the actual goal, though it doesn't tell you the output is (topologically) sorted. `aur-build-order` may be better, but it's not related to `aur-build` apart from the name and that it may be used in conjunction with `aur-build -a`. (__see #383, renamed to `aur-depends`__)
- [X] Clearly mark "helper" scripts that are not meant for interactive use, for example by using a suffix with a double hyphen, e.g. `aur-fetch--git` (__kept with single hyphen, "plumbing"__)

The following were changed or added from the 1.5 release:

- [x] `aur-vercmp`, previously `aurcheck`. `check` doesn't tell you _what_ it checks, so `vercmp` seemed more appropriate and mnemonic with `vercmp(1)`.
- [x] `aur-rpc`, moved out from `aursearch`. A general tool that queries the AUR RPC.
- [x] `aur-srcver`, short for "source version". Pretty much exactly what it does, it queries a bunch of .SRCINFO files to print the version.
- [X] `aur-rfilter`, previously `aursift`. Short for "repo filter". Takes packages from `stdin` (typically AUR packages) and prints valid providers from a specified repository. The "providers" part is what makes the script more than a trivial call to `grep(1)`. (_renamed to `aur-repo-filter`_)
- [x] `aur-pkglist`, previously `aurgrep`. Arguably an implementation detail, although the whole script is centered around `packages.gz` for which there is no real alternative. If `zsync` ever becomes available, this script will likely become obsolete.
- [x] `aur-build-nspawn`, chroot code moved out from `aur-build`. Used to be called `aurbuild_chroot` at some point, which seemed overly generic since the script doesn't work without `systemd-nspawn` at all. (__renamed to `aur-chroot`__)

Any suggestions are welcome, especially for the candidates in the first section.

**Comment by maximbaz** [comment]
`aur-deps-rpc`: How about `aur-rpc-build-order`, which is short for "querying AUR RPC for build order"?

`aur-arrow`: This is the most mysterious one 🙂 This script doesn't seem to be used in any other aurutils script as of the current master, and if it is intended to be piped to `tsort`, why not make this script also do `tsort` as part of it, and also reverse the resulting list? If it did so, I would have suggested to name it `aur-srcdir-build-order`.

**Comment by AladW** [comment]
"child" scripts (with two hyphens) make sense when these describe equivalent ways to proceed in a thin wrapper (with a single hyphen), e.g. `aur-fetch-git` and `aur-fetch-snapshot`. See the pipeline in https://github.com/AladW/aurutils/blob/master/lib/aur-fetch#L64). 

Whether the same holds for `aur-build` is up to debate - `aur-build-nspawn` is essentially an `extra-x86_64-build` replacement with support for local repos, and does cover only a fraction of the `aur-build` functionality.

For `aur-arrow`, I think it's interesting to keep this general and not hardcode `tsort`. For a crazy example, imagine piping the output to `graphviz` for "aurgraph" (see [pacgraph](http://kmkeen.com/pacgraph/)). I guess something like `aur-srcdir-build-order` is descriptive enough, but it's also quite long!

**Comment by maximbaz** [comment]
> "child" scripts (with two hyphens) make sense when these describe equivalent ways to proceed in a thin wrapper (with a single hyphen), e.g. `aur-fetch-git` and `aur-fetch-snapshot`

I see your point, but this convention would have been more obvious if **all** "parent" scripts were named with a single hyphen - for example, right now `aur-deps-rpc` has two, and it can be confused as a "child" script. Even though we will rename it anyway, how do you feel about using a different symbol to indicate "child" scripts that people probably don't want to call directly, for example an underscore? So `aur-fetch_git` and `aur-fetch_snapshot`. That way we don't necessarily have to limit ourselves to think of a name with a single hyphen only, and to me personally this would have been a more clear indication of parent-child relationship.

> For `aur-deps-rpc`, I guess a name like `aur-rpc-build-order` is descriptive enough, but also quite long! 

True that, with bash completion that wouldn't be that big of a deal, but still, all other scripts have much shorter name and this one would stand out. But, it is super descriptive 🙂

> For `aur-arrow`, I think it's interesting to keep this general and not hardcode tsort.

I see. How about `aur-srcdir-deps`? Kind of similar to the old `aur-deps-src`, but more indicative that the script works on a directory of PKGBUILDs, and still explains that it is searching for dependencies in the directory.

**Comment by maximbaz** [comment]
> By the same logic, `aur-search` would be called `aur-rpc-search`!

Yes, although search is only available via RPC, while for deps order we have two scripts, one works on RPC, another on a directory of PKGBUILDs. But given that the second one isn't strictly giving us "order", maybe you are right. How about `aur-deps-order`?

**Comment by maximbaz** [comment]
I'd say it primarily depends on whether you expect people to use this script directly, or it is purely a child for `aur-build`. It would visually help distinguishing which scripts are intended for "public use", and which are just helpers. 

We could even exclude the scripts with two hyphens from the bash completion (I like the two consecutive hyphens by the way).

**Comment by AladW** [comment]
I think `aur-chroot` is suitable for direct use, at least from https://github.com/AladW/aurutils/commit/988da07b88d81fb1d731f679cf58e7f74511ff46

**Comment by maximbaz** [comment]
In that case the name is good 😉

**Comment by rafasc** [comment]
I also feel `aur-chroot` is acceptable.

**Comment by AladW** [comment]
https://github.com/AladW/aurutils/commit/19203f18dd236fb29e387b81b745d02ebcc97f1e. 7 to go :wink:

**Comment by maximbaz** [comment]
For the record, I'm definitely satisfied with `aur-vercmp`, `aur-rpc` and `aur-srcver`; `aur-pkglist` sounds ok too; and I'd probably rename `aur-rfilter` to `aur-repo-filter` for clarity.

TL;DR for already mentioned suggestions that I currently like: 

* two hyphens for `aur-fetch--*`
* `aur-arrow` -> `aur-deps-srcdir`
* `aur-deps-rpc` -> `aur-deps-order`

**Comment by rafasc** [comment]
Regarding the git convention for scripts. Git categorizes it's parts as porcelain/plumbing.  So they're not as dependent on pathname conventions. I think the rule is to keep short and relevant. But they still have some long ones like: `git-get-tar-commit-id`

**Comment by AladW** [comment]
>For the record, I'm definitely satisfied with aur-vercmp, aur-rpc and aur-srcver; aur-pkglist sounds ok too;

I agree, I've put a checkmark for those on the list (still taking any suggestions, of course). 

>I'd probably rename aur-rfilter to aur-repo-filter for clarity.

Sounds good. `r` might mean `reverse` or some other term, after all.

**Comment by AladW** [comment]
>We could even exclude the scripts with two hyphens from the bash completion (I like the two consecutive hyphens by the way).

I was wondering how to exclude scripts like `aur-fetch-git` easily from the completion, and "scripts with a `--` suffix" is well-defined unlike "something that has two or more hyphens somewhere in the word"...

**Comment by AladW** [comment]
I'll have to think a bit more about using `--`, after all `aur-fetch-git` is something you _could_ use directly in a general context, it's just not the most convenient thing to do. Maybe I'm not sure on the exact definition of a "helper", e.g. `gui--askpass` vs. `remote-http`.

Thanks for all the input so far.

**Comment by rafasc** [comment]
I tried to find out something about this convention but did not find anything relevant other than git/git@d7e522cffb2224b91c9c0edde093fbb81a0289b0

In particular, `git-remote-http` is part of the implementation for remote helpers: https://git-scm.com/docs/git-remote-helpers. They are maybe an exception, since they are expecting users to create their own helpers and the `--` would be considered ugly? IDK.

My interpretation is that `--helpers` scripts are used when one tool needs a subtask done (opening a browser, asking for a password, showing a menu in add --interactive, etc). You  never run these helpers by hand. You get their functionality by using their parent commands.

I believe what really defines if a command should be used or not is the commands categorization that can be seen in: [`git(1)`](https://git-scm.com/docs/git) or https://github.com/git/git/blob/master/command-list.txt where porcelain and plumbing are the high and low level interface respectively. What is missing from the list are the ones that are not meant for users.

Using the `git` analog, `aur-fetch` would be porcelain, and `aur-fetch-git` could be either plumbing or a `--heper` depending if we decide users should be using it or not.

One thing git does, by default, is to only auto-complete high-level commands. We could do the same to signal users they should  prefer `aur-fetch -g` over `aur-fetch-git`.

**Comment by AladW** [comment]
Considering `aur-fetch` is a thin wrapper with the bulk of the functionality in `aur-fetch-git` and `aur-fetch-snapshot`, I would say that's it's "plumbing" rather than a "helper". I agree on not using it for completion though, since for interactive use `aur-fetch` is better suited.

Regarding `aur-arrow`, I think it's a good idea to rename it to `aurgraph`. By mapping `pkgname` dependencies to their respective `pkgbases`, it basically creates a directed graph with the package bases as nodes and dependency relations as arrows. It's also mnemonic with `pacgraph` and `pactree`.

I planned on reducing the scope of `aur-deps-rpc`, in that it only recurses the AUR for dependency information and otherwise not process the data. The output would be in the form of `pkgname\tdepends\tpkgbase\tpkgver`. Adapting `aur-fetch` should be simple (note that `aur-fetch -r` currently clones by `pkgname`, so we'd fix a bug in the process). `aur-sync` is more challenging. I have done most of the changes locally but `aur-sync` would need thorough testing.

This way `aur-deps-rpc` both better reflects the name and is a more general (and simple) tool.

**Comment by maximbaz** [comment]
I like the name `aur-graph` 👍 

For `aur-deps-rpc`, notice that this is the only name that explicitly states how it operates (rpc vs source directory), maybe it should just be called `aur-deps`.

**Comment by AladW** [comment]
>For aur-deps-rpc, notice that this is the only name that explicitly states how it operates (rpc vs source directory), maybe it should just be called aur-deps.

Good point, though how about making it slightly longer with `aur-depends`? (compare `depends` array in `PKGBUILD(5)`)?

**Comment by maximbaz** [comment]
One more observation, in light of other "multi-word" scripts such as `srcver`, `vercmp` and `pkglist`, maybe `aur-repo-filter` should become `aur-repofilter`.

Maybe the number of hyphens can even be used in bash completion to exclude "plumbing" scripts.

**Comment by AladW** [comment]
Makes sense, though I find it comparatively long for a single word... maybe the old `aur-sift` is also worth considering.

**Comment by maximbaz** [comment]
`aur-sift` alone is a bit ambiguous, maybe `aur-reposift`?

**Comment by rafasc** [comment]
Agree that `aur-graph` sounds good.

I would try to stay away from using the number of non-consecutive hyphens to categorize commands. I believe it's too generic of a pattern and it would restrict us in the future if we can't come up with a short and good name for a script. Some commands suggestions from this thread look awful in their "Hyphenless Multi-word" form. (depsorder, depssrcdir, rpcbuildorder) just to give some examples.

I would say the best route is to explicitly define the categories, like git.

Making it available on something like [`aur-help`](
https://github.com/rafasc/aurutils/blob/wip_move_to_getopt/lib/aur-help#L247) where it could be easily called from both bash and zsh compeltions with `aur help --cmd-list [basic|core|usrcmds]` (categories names up for discussion)

Holdin on a bit longer on this branch, now that we are renaming a bunch of stuff and pacman 5.1 is slushy freezed (need to look into parseopts.sh).

**Comment by AladW** [comment]
In that case I think we can leave the `aur-repo-filter` name for now.

For the category names, perhaps the more easily readable `user` instead of `usrcmds`? 

Otherwise `basic` and `core` sound decent, though slightly vague, i.e. there "basic" does not necessarily imply "for interactive use" to me.

**Comment by rafasc** [comment]
That can be discussed on its PR. I think we can close this. We can always
reopen it if needed.

**Comment by AladW** [comment]
:+1:
